### PR TITLE
Added symmetry breaking clauses

### PR DESCRIPTION
**Pull request by vint-1**
- BFS symmetry breaking clauses based on Ulyantsev 2016
- Added some test cases

### CODE REVIEW

File: dfa_identify/encoding.py [hunk 1]
```diff
@@ -44,22 +49,75 @@ class Codec:
...
+    counts: list[int] = attr.ib()
+    @counts.default
+    def counts_default(self):
+        """
+        Compute number of variables of each type.
+        Used for calculating offsets for decoding and encoding
+        Order: z, x, y, p, t, m
+        """
+        return [
+            self.n_colors,
```
Comment (mvcisback) [inline dfa_identify/encoding.py:63]: Maybe break up into 6 properties for the various counts?

File: dfa_identify/encoding.py [hunk 2]
```diff
@@ -44,22 +49,75 @@ class Codec:
...
     def parent_relation(self, token: Any, color1: int, color2: int) -> int:
+        """ Literature refers to these variables as y """
+        assert (color1 >= 0) and (color2 >= 0), "colors must be nonnegative"
+        assert (color2 < self.n_colors) and (color1 < self.n_colors), "colors must be smaller than n_colors"
+        assert (token < self.n_tokens) and (token >= 0), "token must be nonnegative and smaller than n_tokens"
         a = self.n_colors
         b = a**2
-        c = 1 + self.n_colors * (1 + self.n_nodes)
-        return color1 + a * color2 + b * token + c
+        # c = 1 + self.n_colors * (1 + self.n_nodes)
```
Comment (mvcisback) [inline dfa_identify/encoding.py:95]: please delete dead code :)

File: dfa_identify/encoding.py [hunk 3]
```diff
@@ -44,22 +49,75 @@ class Codec:
...
+        # return color1 + a * color2 + b * token + c
+        return sum(self.counts[:2]) + 1 + color1 + a * color2 + b * token
+
+    def enumeration_parent(self, color1: int, color2: int) -> int:
+        """ Literature refers to these variables as p
+        Note: here we use p_{i,j} rather than p_{j,i} """
+        assert (color1 < color2), "color1 must be smaller"
```
Comment (mvcisback) [inline dfa_identify/encoding.py:102]: Maybe create a validation decorator?

File: dfa_identify/encoding.py [hunk 4]
```diff
@@ -44,22 +49,75 @@ class Codec:
     n_nodes: int
     n_colors: int
     n_tokens: int
+    symm_mode: SymmBreak = attr.ib(default = SymmBreak.CLIQUE)
+
+    counts: list[int] = attr.ib()
```
Comment (mvcisback) [inline dfa_identify/encoding.py:54]: Maybe make this s tuple so you can't accidentally change late.

File: dfa_identify/encoding.py [hunk 5]
```diff
@@ -81,23 +139,29 @@ def decode(self, lit: int) -> Var:
 # ================= Clause Generator =====================
 
 
-def dfa_id_encodings(apta: APTA) -> Iterable[Clauses]:
+def dfa_id_encodings(apta: APTA, symm_mode: SymmBreak = SymmBreak.CLIQUE) -> Iterable[Clauses]:
     cgraph = apta.consistency_graph()
     clique = max_clique(cgraph)
```
Comment (mvcisback) [inline dfa_identify/encoding.py:144]: `clique = .. if symm_mode is not None else None`

File: dfa_identify/encoding.py [hunk 6]
```diff
@@ -81,23 +139,29 @@ def decode(self, lit: int) -> Var:
...
 
     for n_colors in range(len(clique), len(apta.nodes) + 1):
-        codec = Codec.from_apta(apta, n_colors)
-        yield codec, list(encode_dfa_id(apta, codec, clique, cgraph))
-
+        codec = Codec.from_apta(apta, n_colors, symm_mode = symm_mode)
+        if symm_mode == SymmBreak.NONE:
```
Comment (mvcisback) [inline dfa_identify/encoding.py:148]: Above proposal means you don't need to case split here.

File: dfa_identify/encoding.py [hunk 7]
```diff
@@ -81,23 +139,29 @@ def decode(self, lit: int) -> Var:
...
+            yield codec, list(encode_dfa_id(apta, codec, cgraph, clique))
 
-def encode_dfa_id(apta, codec, clique, cgraph):
+def encode_dfa_id(apta, codec, cgraph, clique = None):
     # Clauses from Table 1.                                      rows
     yield from onehot_color_clauses(codec)                     # 1, 5
     yield from partition_by_accepting_clauses(codec, apta)     # 2
     yield from colors_parent_rel_coupling_clauses(codec, apta) # 3, 7
     yield from onehot_parent_relation_clauses(codec)           # 4, 6
     yield from determination_conflicts(codec, cgraph)          # 8
```
Comment (mvcisback) [inline dfa_identify/encoding.py:159]: Could we add an option for turning this on and off as well?

File: dfa_identify/encoding.py [hunk 8]
```diff
@@ -160,8 +223,53 @@ def determination_conflicts(codec: Codec, cgraph: nx.Graph) -> Clauses:
...
+    See Ulyantsev 2016 
+    """
+    yield [codec.color_node(0,0)] # Ensures start vertex is 0 - not listed in Ulyantsev
+    for color2 in range(codec.n_colors):
+        if color2 > 0:
+            yield [codec.enumeration_parent(color1, color2) for color1 in range(color2)] # 4
+        for color1 in range(color2):
+            p = codec.enumeration_parent(color1, color2)
+            t = codec.transition_relation(color1, color2)
+            m = lambda l: codec.enumeration_label(l, color2)
```
Comment (mvcisback) [inline dfa_identify/encoding.py:241]: `functools.partial` to help track down where this function comes from later.

File: tests/test_identify.py [hunk 9]
```diff
@@ -1,26 +1,160 @@
...
 
-    for x in accepting:
-        assert my_dfa.label(x)
+        for i in range(50):
+            self.solve_and_check(self.problem5, symm_mode = symm)
 
-    for x in rejecting:
-        assert not my_dfa.label(x)
+if __name__ == '__main__':
+    unittest.main()
```
Comment (mvcisback) [inline tests/test_identify.py:160]: Currently this project uses `pytest` is there a reason you want to introduce these?

File: tests/test_identify.py [hunk 10]
```diff
@@ -1,26 +1,160 @@
...
+                x += str(np.random.choice(alphabet))
+            if f(x):
+                positives.add(x)
+            else:
+                negatives.add(x)
+        return list(positives), list(negatives)
 
-    accepting = [[0], [0, 'z', 0, 0], ['z', 'z']]
-    rejecting = [[0, 'z', 'z'], ['z']]
+    def generate_problem(self, n, input_sz, bypass = False):
```
Comment (mvcisback) [inline tests/test_identify.py:55]: To make this reproducible, lets use https://hypothesis.readthedocs.io/en/latest

File: tests/test_identify.py [hunk 11]
```diff
@@ -1,26 +1,160 @@
...
+import time
+import os
+import unittest
+import attr
 
+@attr.s(auto_detect=True, auto_attribs=True, frozen=True)
+class IdProblem:
+    accepting: list[str]
+    rejecting: list[str]
+    size: int # min DFA size
```
Comment (mvcisback) [inline tests/test_identify.py:14]: Maybe `hypothesis.just` decorator?

File: tests/test_encoding.py [hunk 12]
```diff
@@ -1,80 +1,103 @@
...
+            assert var.parent_color == c1
+            assert var.node_color == c2
+            assert var.token == t
 
-    assert len(lits) == 9 * 4  # Check bijection.
-    assert max(lits) == 3 + 3 * 10 + 9 * 4
+        assert len(lits) == 9 * 4  # Check bijection.
+        assert max(lits) == 3 + 3 * 10 + 9 * 4
 
+    def test_symm_break(self):
```
Comment (mvcisback) [inline tests/test_encoding.py:70]: Also, can we also test if symmetry break is correct by checking that there is at most 1 model.

We could do this a number of ways, but my proposal would be to take the codec and solve for the model as in the identify function.

Then assert that that particular model is not allowed by adding a clause with the negated literals.

`diff_model = [-lit for lit in model]`

Adding that clause to the SAT solver should make the instance unsat.

File: tests/test_identify.py [hunk 13]
```diff
@@ -1,26 +1,160 @@
...
+            t0 = time.time()
+            dfa1 = self.solve_and_check(problems[n], symm_mode=symm)
+            elapsed = time.time() - t0
+            
+            print("solved n = {} with no symmetry-breaking in {:.4g}s".format(n,elapsed))
+            write_dot(dfa1, os.path.join(self.outdir,"longtest_n-{}_symm-{}.dot".format(n,symm)))
+        
+        symm = SymmBreak.CLIQUE
+        print("Running tests for clique-based symmetry breaking...")
+        for n in range(start, end, step):
```
Comment (mvcisback) [inline tests/test_identify.py:123]: Can you break out the benchmarking into a different file? If we're checking for performance regressions, we can also set a deadline in hypothesis.

File: tests/test_identify.py [hunk 14]
```diff
@@ -1,26 +1,160 @@
...
+            write_dot(dfa1, os.path.join(self.outdir,"longtest_n-{}_symm-{}.dot".format(n,symm)))
+
+    def test_identify_repeatedly(self):
+
+        symm = SymmBreak.BFS
+
+        for i in range(50):
+            self.solve_and_check(self.problem1, symm_mode = symm)
+
+        for i in range(50):
```
Comment (mvcisback) [inline tests/test_identify.py:150]: This is just checking determinism right?

DFA is getting an equiv check which we could use. Alternatively, we could check that the SAT solution is always the same model.

Comment (mvcisback) [reply tests/test_identify.py:150]: 50 is also probably too much outside of a benchmark. 3 should be sufficient unless your experience says otherwise.

**Review by mvcisback — CHANGES_REQUESTED** [review]
Hi,

Looks really good. Great job!

I have some style things that I'd change before merging. If you don't mind taking care of them that would be really helpful.

For testing, this and most of my projects use `pytest`. I usually run:

`poetry run pytest --flake8`    (its in the drone.yml which is run for continuous integration)

File: dfa_identify/encoding.py (hunk 1 above)
Comment (vint-1) [reply dfa_identify/encoding.py:64]: It's easier to have it as a tuple, because we need to sum the counts later

**Pull request merged**